#Перевірка статиcтичних гіпотез. Z-test.



0. Зчитайте дані з `data.csv` у змінну data, яка має тип pandas.DataFrame. Ми будемо далі працювати з цією змінною.

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats

# Налаштування для візуалізації
plt.rcParams['figure.figsize'] = (12, 7)
plt.rcParams['font.size'] = 12
data = pd.read_csv('data.csv')

Запустіть код нижче. Для коректної робити всіх подальших методів дані мають бути у вигляді одновимірного `numpy.ndarray` вектора та мати розмірність (100,). Така розмірність означає, що дані - одновимірні (колонка), якщо б розмірність була (100, 1), то дані сприймаються методами як двовимірні (таблиця), хоча для нас виглядати можуть так само.

In [2]:
data = data.data.values
data.shape

(100,)

In [3]:
print(data)

[17.49945292 19.66239904  7.18282343 29.84162485  9.23938649 14.94951581
 23.0172885  12.52827148 13.65228669 14.54595431 23.30872427 33.75324808
 20.24923636 13.29244733 23.23434992 16.4230418  19.88521702 27.05000732
 15.5127743  20.05415151 14.73135264 19.06139498 21.53942271 14.06732571
 17.9670682  18.58289581 16.17406993 12.87432628 11.47269664 19.07902883
 18.38565824 33.38820073  5.39139454 20.67635903 22.22266722 28.15780318
 23.01114324 14.93471778 20.00005857 23.25411543 18.11895082 24.62607043
  8.79145607 30.387108   28.80606806 17.98593597 23.66804468 20.28782355
 15.02518827 20.52626131 26.00219532 17.71344489 17.74598346 19.55317542
 22.60097798 27.67027538 16.19192417 23.05037746 21.29669604  8.84832568
 17.48410111 19.20602661 19.76257856 21.9560206   7.75806171 20.27753314
 15.93394654 11.36336584 23.14577858 24.41167746 16.08049839 25.05473769
 17.71090111 20.39893405 13.40756632 29.50692234  4.04330326 19.45128426
 24.17071763  7.79920072 18.86318441 19.53668801 24

**Завдання 1**. (7 балів)
Зчитані дані - це сума покупок в доларах наших покупців на сайті протягом липня 2025 року.

До цього місяця, ми мали продажі в середньому на чек $20.

Необхідно зʼясувати, чи вийшло у нас статистично значущо **підвищити** середній чек за липень 2025?

Аби дати відповідь - ми проведемо z-test на рівні значущості $\alpha=0.05$ двома способами. В цьому завданні ми будемо виконувати обчислення "вручну" з використанням бібліотек numpy та scipy.stats подібно до прикладу в лекції. Для цього виконайте наступні 7 кроків. Правильне виконання кожного з кроків оцінюється в 1 бал.

1. Запишіть параметри задачі у змінні Python:
    - `sample_mean` - $\bar{x}$, середнє значення в вибірці
    - `population_mean` - $\mu_0$, середнє значення в популяції (тобто те, з яким ми порівнюємо середнє вибіркове значення)
    - `population_std` - $\sigma$, вибіркове стандартне відхилення, яке ми вважаємо, що є рівним ст. відх. популяції, адже маємо достатньо велику вибірку.
    - `sample_size` - $n$, розмір вибірки
    - `alpha` - $\alpha$ рівень значущості
    
    **Увага!** Для обчислення стандартного відхилення маємо скористатись функцією `np.std(your_dataframe, ddof=1)`. Чому так - розбираємо в лекції про t-test.


2. Визначте для цієї задачі:
    - якою є гіпотеза $H_0$
    - якою є альтернативна гіпотеза $H_a$
    - з яким типом тесту ми маємо справу - лівосторонній, правосторонній чи двосторонній.

3. Обчисліть стандартну помилку SE.
4. Розрахуйте z-статистику.
5. Знайдіть критичне z-значення з допомогою бібліотеки stats.
6. Обчисліть p-value з допомогою бібліотеки stats.
7. Прийміть рішення, чи відхиляєте ви гіпотезу $H_0$. Для прийняття рішення зробіть і порівняння z-статистики з критичним значенням, і проаналізуйте p-value.

In [4]:
sample_mean = np.mean(data)
print(f"Середнє значення в вибірці: x̄ = {sample_mean:.3f}")
population_mean = 20
print(f"Середнє значення в популяції: μ₀ = {population_mean}")
population_std = np.std(data, ddof=1)
print(f"Вибіркове стандартне відхилення: σ = {population_std:.3f}")
sample_size = len(data)
print(f"Розмір вибірки: n = {sample_size}")
alpha = 0.05
print(f"Рівень значущості: α = {alpha}")

Середнє значення в вибірці: x̄ = 19.378
Середнє значення в популяції: μ₀ = 20
Вибіркове стандартне відхилення: σ = 6.254
Розмір вибірки: n = 100
Рівень значущості: α = 0.05


**Гіпотези:**   

H₀: μ = 20 (продажі в середньому на чек = $20 до липня місяця)  
H₁: μ > 20 (продажі в середньому на чек збільшилися у липні місяці)

**Тип тесту:** правосторонній (one-tailed)  

Стверджується, що продажі в середньому на чек в липні місяці **збільшилися** (не просто відрізняються), тому перевіряємо, чи продажі в середньому на чек в липні місяці **більші**, а не просто відрізняються.

**Стандартна помилка:**

σ = 6.254 - вибіркове стандартне відхилення, яке вважаємо, що є рівним стандартному відхиленню популяції

SE = σ/√n = 6.254/√100 ≈ 0.6254 - стандартна помилка

**Z-статистика:**

x̄ = 19.378 - середнє значення в вибірці

μ₀ = 20 - середнє значення в популяції

SE ≈ 0.6254 - стандартна помилка

z = (x̄ - μ₀) / SE = (19.378 - 20) / 0.6254 ≈ -0.995 - z-статистичне (розрахункове)

z' = 1.645 - z-критичне


In [5]:
# Формула: SE = σ / √n
standard_error = population_std / np.sqrt(sample_size)

print("СТАНДАРТНА ПОМИЛКА:")
print("Формула: SE = σ / √n")
print(f"SE = {population_std:.3f} / √{sample_size}")
print(f"SE = {population_std:.3f} / {np.sqrt(sample_size):.3f}")
print(f"SE = {standard_error:.3f}")
print()
print("Інтерпретація: В середньому, вибіркове середнє відхиляється")
print(f"від справжнього середнього на ±{standard_error:.3f}$")

СТАНДАРТНА ПОМИЛКА:
Формула: SE = σ / √n
SE = 6.254 / √100
SE = 6.254 / 10.000
SE = 0.625

Інтерпретація: В середньому, вибіркове середнє відхиляється
від справжнього середнього на ±0.625$


In [6]:
# Формула: z = (x̄ - μ₀) / SE
z_statistic = (sample_mean - population_mean) / standard_error

print("Z-СТАТИСТИКА:")
print("Формула: z = (x̄ - μ₀) / SE")
print(f"z = ({sample_mean:.3f} - {population_mean}) / {standard_error:.3f}")
print(f"z = {(sample_mean - population_mean):.3f} / {standard_error:.3f}")
print(f"z = {z_statistic:.3f}")
print()
print("Інтерпретація:")
print(f"Наше вибіркове середнє відхиляється на {z_statistic:.3f} стандартних")
print("помилок від гіпотетичного середнього популяції.")
print()
if z_statistic > 2:
    print("Це досить велике відхилення. (більше 2 стандартних помилок)")
elif z_statistic > 1:
    print("Це помірне відхилення (між 1 та 2 стандартними помилками)")
else:
    print("Це невелике відхилення (менше 1 стандартної помилки)")

Z-СТАТИСТИКА:
Формула: z = (x̄ - μ₀) / SE
z = (19.378 - 20) / 0.625
z = -0.622 / 0.625
z = -0.995

Інтерпретація:
Наше вибіркове середнє відхиляється на -0.995 стандартних
помилок від гіпотетичного середнього популяції.

Це невелике відхилення (менше 1 стандартної помилки)


In [7]:
# Для правостороннього тесту з α = 0.05
z_critical = stats.norm.ppf(1 - alpha)

print("КРИТИЧНЕ ЗНАЧЕННЯ:")
print(f"Для рівня значущості α = {alpha} (правосторонній тест)")
print(f"Критичне z-значення = {z_critical:.3f}")
print()
print("Що це означає?")
if z_statistic > z_critical:
    print(f"z-статистичне = {z_statistic:.3f} > z-критичне = {z_critical:.3f}")
    print("Результат настільки незвичайний,\nщо ми не віримо у випадковість за гіпотезою Н0.")
    print("Гіпотезу Н0 відхилено.")
    print(f"Ймовірність отримати z > {z_critical:.3f} випадково = {alpha*100}%")
else:
    print(f"z-статистичне = {z_statistic:.3f} < z-критичне = {z_critical:.3f}")
    print("Гіпотезу Н0 не відхилено.")
    print(f"Ймовірність отримати z > {z_critical:.3f} випадково = {alpha*100}%")

КРИТИЧНЕ ЗНАЧЕННЯ:
Для рівня значущості α = 0.05 (правосторонній тест)
Критичне z-значення = 1.645

Що це означає?
z-статистичне = -0.995 < z-критичне = 1.645
Гіпотезу Н0 не відхилено.
Ймовірність отримати z > 1.645 випадково = 5.0%


In [8]:
# Для правостороннього тесту: p-value = P(Z > z_statistic)
p_value = 1 - stats.norm.cdf(z_statistic)

print("P-VALUE:")
print(f"p-value = {p_value:.6f}")
print()
print("Що це означає?")
print(f"Якщо H₀ правильна (продажі в середньому на чек = $20), то ймовірність")
print(f"випадково отримати продаж в середньому на чек ≥ {sample_mean:.3f} становить {p_value:.6f}")
print(f"або приблизно {p_value*100:.4f}%")
print()
if p_value < 0.001:
    print("Це НАДЗВИЧАЙНО малоймовірно! (менше 0.1%)")
elif p_value < 0.01:
    print("Це дуже малоймовірно! (менше 1%)")
elif p_value < 0.05:
    print("Це малоймовірно! (менше 5%)")
else:
    print("Це цілком можливо випадково")

P-VALUE:
p-value = 0.840216

Що це означає?
Якщо H₀ правильна (продажі в середньому на чек = $20), то ймовірність
випадково отримати продаж в середньому на чек ≥ 19.378 становить 0.840216
або приблизно 84.0216%

Це цілком можливо випадково


In [9]:
print("МЕТОД 1: Порівняння z-статистики з критичним значенням")
print(f"z-статистика = {z_statistic:.3f}")
print(f"Критичне значення = {z_critical:.3f}")
print(f"Порівняння: {z_statistic:.3f} {'>' if z_statistic > z_critical else '≤'} {z_critical:.3f}")

if z_statistic > z_critical:
    print("Висновок: z-статистика ПЕРЕВИЩУЄ критичне значення")
    decision1 = "ВІДХИЛЯЄМО H₀"
else:
    print("Висновок: z-статистика НЕ перевищує критичне значення")
    decision1 = "НЕ ВІДХИЛЯЄМО H₀"

print(f"Рішення: {decision1}")

МЕТОД 1: Порівняння z-статистики з критичним значенням
z-статистика = -0.995
Критичне значення = 1.645
Порівняння: -0.995 ≤ 1.645
Висновок: z-статистика НЕ перевищує критичне значення
Рішення: НЕ ВІДХИЛЯЄМО H₀


In [10]:
print("МЕТОД 2: Порівняння p-value з рівнем значущості")
print(f"p-value = {p_value:.6f}")
print(f"Рівень значущості α = {alpha}")
print(f"Порівняння: {p_value:.6f} {'<' if p_value < alpha else '≥'} {alpha}")

if p_value < alpha:
    print("Висновок: p-value МЕНШЕ за рівень значущості")
    decision2 = "ВІДХИЛЯЄМО H₀"
else:
    print("Висновок: p-value НЕ менше за рівень значущості")
    decision2 = "НЕ ВІДХИЛЯЄМО H₀"

print(f"Рішення: {decision2}")

МЕТОД 2: Порівняння p-value з рівнем значущості
p-value = 0.840216
Рівень значущості α = 0.05
Порівняння: 0.840216 ≥ 0.05
Висновок: p-value НЕ менше за рівень значущості
Рішення: НЕ ВІДХИЛЯЄМО H₀


In [11]:
print("ФІНАЛЬНИЙ ВИСНОВОК:")
if decision2 == "ВІДХИЛЯЄМО H₀":
    print("✅ Твердження ПІДТВЕРДЖУЄТЬСЯ!")
    print("   Вдалося підвищити середній чек за липень 2025 на рівні значущості 5%")
    print(f"   Середній чек за липень 2025 ({sample_mean}) статистично значуще")
    print(f"   перевищує середній чек за попередні місяці ({population_mean})")
else:
    print("❌ Твердження НЕ ПІДТВЕРДЖУЄТЬСЯ")
    print("   Немає достатніх статистичних доказів, що середній чек за липень 2025")
    print("   вдалося статистично значущо підвищити в порівнянні з попередніми місяцями")

ФІНАЛЬНИЙ ВИСНОВОК:
❌ Твердження НЕ ПІДТВЕРДЖУЄТЬСЯ
   Немає достатніх статистичних доказів, що середній чек за липень 2025
   вдалося статистично значущо підвищити в порівнянні з попередніми місяцями


**Завдання 2.** (3 бали)
Виконайте обчислення z-test з використанням бібліотеки statsmodels.

Отримайте z-статистику та р-значення.

Виведіть p-значення та зробіть висновок, чи ми маємо достатньо доказів, аби стверджувати, що середній чек зріс.

Чи зійшлись значення z-статистику та р-значення в цьому завданні з попередніми обчисленнями?

In [12]:
from statsmodels.stats import weightstats as stests

print(f"Дані:")
print(f"  Кількість сум покупок протягом липня: {len(data)}")
print(f"  Середнє сума покупки протягом липня: {np.mean(data):.3f}")
print(f"  Стандартне відхилення: {np.std(data, ddof=1):.3f}")
print(f"  Мінімальна сума покупки протягом липня: {np.min(data):.0f}")
print(f"  Максимальне сума покупки протягом липня: {np.max(data):.0f}")

Дані:
  Кількість сум покупок протягом липня: 100
  Середнє сума покупки протягом липня: 19.378
  Стандартне відхилення: 6.254
  Мінімальна сума покупки протягом липня: 4
  Максимальне сума покупки протягом липня: 34


In [13]:
print("ВИКОНАННЯ Z-ТЕСТУ:")

# Виконуємо правосторонній z-тест
z_stat_sm, p_val_sm = stests.ztest(data,           # наші дані
                                   value=population_mean,  # μ₀ = 100
                                   alternative='larger')   # H₁: μ > 100

print(f"Результати z-тесту:")
print(f"  Z-статистика: {z_stat_sm:.3f}")
print(f"  P-value: {p_val_sm:.6f}")
print()

# Інтерпретація результатів
print("ІНТЕРПРЕТАЦІЯ:")

if p_val_sm < alpha:
    print(f"✅ ВІДХИЛЯЄМО H₀")
    print(f"   p-value ({p_val_sm:.6f}) < α ({alpha})")
    print("   Висновок: Продажі в середньому на чек в липні статистично значуще вищі за попередні місяці")
else:
    print(f"❌ НЕ ВІДХИЛЯЄМО H₀")
    print(f"   p-value ({p_val_sm:.6f}) ≥ α ({alpha})")
    print("   Висновок: Немає достатніх доказів про вищі продажі в середньому на чек в липні 2025")

# Порівняння з ручними розрахунками
print()
print("ПОРІВНЯННЯ З РУЧНИМИ РОЗРАХУНКАМИ:")
print(f"Z-статистика:")
print(f"  Ручний розрахунок: {z_statistic:.3f}")
print(f"  Statsmodels:       {z_stat_sm:.3f}")
print(f"  Різниця:           {abs(z_statistic - z_stat_sm):.3f}")
print(f"P-value:")
print(f"  Ручний розрахунок: {p_value:.3f}")
print(f"  Statsmodels:       {p_val_sm:.3f}")
print(f"  Різниця:           {abs(p_value - p_val_sm):.3f}")
print()

ВИКОНАННЯ Z-ТЕСТУ:
Результати z-тесту:
  Z-статистика: -0.995
  P-value: 0.840216

ІНТЕРПРЕТАЦІЯ:
❌ НЕ ВІДХИЛЯЄМО H₀
   p-value (0.840216) ≥ α (0.05)
   Висновок: Немає достатніх доказів про вищі продажі в середньому на чек в липні 2025

ПОРІВНЯННЯ З РУЧНИМИ РОЗРАХУНКАМИ:
Z-статистика:
  Ручний розрахунок: -0.995
  Statsmodels:       -0.995
  Різниця:           0.000
P-value:
  Ручний розрахунок: 0.840
  Statsmodels:       0.840
  Різниця:           0.000

